
###### standard curation of the input ######
# construct a conversion table between flybase id and gene name 
# construct the meme matrix for different CrebA motifs 

# build the flybase dictionary 
rule build_flybase_dict: 
    input: 
        "../input/reference_genome/dmel-all-r6.33.gtf"
    output: 
        "../input/flybase_gene_conversion/conversion_tab.csv"
    shell: 
        "python3 build_flybase_dictionary.py"
    
# build the meme matrix 
rule convert_chen_to_meme: 
    input: 
        "../input/previous_CrebA_motifs/AllCrebAMotifs_probabilitymatricies.txt"
    output: 
        "../input/previous_CrebA_motifs/CrebAMotifs_meme.meme"
    shell: 
        "python3 chen_to_meme.py"

###### gather the promoter regions and +/- 1kb regions surrounding the TSS of genes #######

# get the promoter regions of the genes 
rule get_promoter_regions: 
    input: 
        "../input/reference_genome/dmel-all-r6.33.gtf", 
        "../input/reference_genome/dmel-all-chromosome-r6.33.fasta"
    output:
        "../output/tss_table/promoter_regions.bed"
    shell: 
        "python3 get_promoter.py"

##### trim peak regions to be 250 +/- around the peak and get the intersecting peaks ######
# trim the peak regions to be 250 +/- around the peak 
rule trim_range_peaks: 
    input: 
        "../input/GEO_processed/GSM4213092_oregon_CrebA_peaks.narrowPeak",
        "../input/GEO_processed/GSM4213094_fkh_CrebA_peaks.narrowPeak",
        "../input/GEO_processed/GSM4213096_sage_CrebA_peaks.narrowPeak"
    output: 
        "../output/range_peak_regions/CrebA_oregon.narrowPeak", 
        "../output/range_peak_regions/CrebA_fkh.narrowPeak",
        "../output/range_peak_regions/CrebA_sage.narrowPeak"
    shell: 
        """
        mkdir -p ../output/range_peak_regions/
        python trim_peaks_range.py --range 250 --bed {input[0]} --out {output[0]}
        python trim_peaks_range.py --range 250 --bed {input[1]} --out {output[1]}
        python trim_peaks_range.py --range 250 --bed {input[2]} --out {output[2]}
        """

# find sage intersecting fkh ChIP-seq 
rule find_fkh_sage_intersect_peaks: 
    input: 
        "../output/range_peak_regions/CrebA_fkh.narrowPeak", 
        "../output/range_peak_regions/CrebA_sage.narrowPeak"
    output: 
        "../output/range_peak_regions/fkh_sage_intersect.narrowPeak"
    shell: 
        """
        bedtools intersect -a {input[1]} -b {input[0]} -f 0.5 -u -wa > {output}  
        """

##### match the peaks to a gene ######
# match the peak to a gene 
rule match_peak_gene: 
    input: 
        "../output/range_peak_regions/fkh_sage_intersect.narrowPeak"
    output: 
        "../output/match_nearest_gene/fkh_sage_intersect_genes_1000.csv",
        "../output/match_nearest_gene/fkh_sage_intersect_genes_2000.csv"
    shell: 
        """
        python3 match_nearest_region_to_genes.py --bed {input[0]} --maxLength 1000 --out {output[0]}
        python3 match_nearest_region_to_genes.py --bed {input[0]} --maxLength 2000 --out {output[1]}
        """

# find the SPCGs that are bound by CrebA 
rule find_bound_SPCGs: 
    input: 
        "../output/match_nearest_gene/fkh_sage_intersect_genes_1000.csv",
        "../output/match_nearest_gene/fkh_sage_intersect_genes_2000.csv"
    output: 
        "../output/find_bound_SPCGs/spcg_match_1000.csv", 
        "../output/find_bound_SPCGs/spcg_match_2000.csv"
    shell: 
        """
        mkdir -p ../output/find_bound_SPCGs
        python3 find_bound_SPCGs.py --input {input[0]} --output {output[0]}
        python3 find_bound_SPCGs.py --input {input[1]} --output {output[1]}
        """

# the rule to find bound DE genes 
rule find_bound_DE_genes: 
    input: 
        "../output/match_nearest_gene/fkh_sage_intersect_genes_1000.csv"
    output: 
    shell: 
        """
        python3 find_bound_DE_genes.py
        Rscript enrichment_categorized_DE_genes.R
        """

rule plot_intersect_bound_genes: 
    input: 
        "../output/find_bound_DE_genes/down_DE.csv",
        "../output/find_bound_DE_genes/up_DE.csv",
        "../output/find_bound_SPCGs/spcg_match.csv",
        "../output/match_nearest_gene/fkh_sage_intersect_genes_1000.csv"
    output: 
        "../output/bound_spcg_DE_venn/bound_SPCGs_down_venn.png", 
        "../output/bound_spcg_DE_venn/bound_SPCGs_scdown_venn.png", 
        "../output/bound_spcg_DE_venn/bound_SPCGs_up_venn.png", 
        "../output/bound_spcg_DE_venn/bound_SPCGs_scup_venn.png", 
        "../output/bound_spcg_DE_venn/bound_down_up_venn.png",
    shell: 
        """
        python3 plot_intersect_bound_genes.py
        """

rule plot_distance_from_tss: 
    input: 
        '../output/match_nearest_gene/fkh_sage_intersect_genes.csv',
        "../output/tss_table/regulatory_regions_table.txt",
        "../../analysis/results/v19/DE_genes_early_crebA_wt/Salivary Gland/mut_DE_genes.csv",
        "../output/find_bound_DE_genes/down_DE.csv",
        "../output/find_bound_DE_genes/up_DE.csv"
    output: 
        directory('../output/plot_location_crebA_binding')
    shell: 
        """
        Rscript plot_distance_crebA.R
        """

